src/core/Internal.ml

```diff
@@ -1661,21 +1699,26 @@ module Make(Plugin : PLUGIN)
     | Solver_intf.Eval l ->
       let a = mk_atom st f in
       enqueue_semantic st a l
-    | Solver_intf.Consequence (causes, proof) ->
-      let l = List.rev_map (fun f -> Atom.neg @@ mk_atom st f) causes in
-      if List.exists Atom.is_true l then (
-        invalid_argf "slice.acts_propagate: Consequence should contain only true literals"
-      );
+    | Solver_intf.Consequence mk_expl ->
       let p = mk_atom st f in
       if Atom.is_true p then ()
       else if Atom.is_false p then (
+        let lits, proof = mk_expl() in
+        let l = List.rev_map (fun f -> Atom.neg @@ mk_atom st f) lits in
+        if List.exists Atom.is_true l then (
+          invalid_argf "slice.acts_propagate: Consequence should contain only true literals"
+        );
         let c = Clause.make_removable (p :: l) (Lemma proof) in
         raise_notrace (Th_conflict c)
       ) else (
         insert_var_order st p.var;
-        let level = List.fold_left (fun acc a -> max acc a.var.v_level) 0 l in
-        let c = Clause.make_removable (p :: l) (Lemma proof) in
-        enqueue_bool st p ~level (Bcp c)
+        let c = lazy (
+          let lits, proof = mk_expl () in
```

Comment:
Isn't this use of lazy potentially problematic if the theory has a mutable state ? If not, maybe add a comment explaining why there can't be problems.

Reply:
it's potentially problematic, but if your theory doesn't have proper backtracking hygiene nothing prevents you from computing explanations eagerly.

Reply:
- no it's not tested, I don't see a simple way of testing except by using msat in a real project
- doing propagations late is quite weird to me, it's a bit nonsensical in CDCL(T) and still strange in mcSAT, I think. This model is simple and should work with low overhead (given that most propagations won't lead to a conflict, as I read in at least one paper on lazy SMT). If you want to propagate at a much lower level, the simplest is probably to add a non-persistent lemma.